frontend/kubernetes/plugin.go

```diff
@@ -1508,3 +1542,178 @@ func (p *Plugin) updatePVCSize(
+			"the source is not a PVC")
+		return
+	}
+
+	log.WithFields(log.Fields{
+		"volumeSnapshot": volSnap.Name,
+		"namespace":      volSnap.Namespace,
+		"sourceKind":     volSnap.Spec.Source.Kind,
+		"sourceName":     volSnap.Spec.Source.Name,
+	}).Debug("Kubernetes frontend got notified of a VolumeSnapshot CR.")
+
+	// Validate the source PVC
+	// 1. PVC and VolumeSnapshotCR must be in the same namespace
+	sourcePVC := volSnap.Spec.Source.Name
+	k8sClient, err := p.getNamespacedKubeClient(&p.kubeConfig, volSnap.Namespace)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"namespace": volSnap.Namespace,
+			"error":     err.Error(),
+		}).Error("Kubernetes frontend couldn't create a client to namespace!")
+		return
+	}
+	claim, err := k8sClient.GetPVC(sourcePVC, metav1.GetOptions{})
+	if err != nil {
+		err = fmt.Errorf("snapshotting a PVC requires the PVC to be in the same namespace as the VolumeSnapshot CR")
+		log.WithFields(log.Fields{
+			"volumeSnapshot": volSnap.Name,
+			"sourcePVC":      sourcePVC,
+		}).Errorf("Kubernetes frontend detected an invalid configuration "+
+			"for creating snapshot from a PVC: %v", err.Error())
+		return
+	}
+
+	// 2. Trident should be set as the provisioner for the PVC
+	provisioner := getClaimProvisioner(claim)
+	if provisioner != AnnOrchestrator &&
+		GetPersistentVolumeClaimClass(claim) != "" {
+		log.WithFields(log.Fields{
+			"volumeSnapshot":    volSnap.Name,
+			"sourcePVC":         claim.Name,
+			"sourceProvisioner": provisioner,
+		}).Debugf("Kubernetes frontend ignored this VolumeSnapshot as the source PVC is not "+
+			"tagged with %s as the storage provisioner!", AnnOrchestrator)
+		return
+	}
+
+	// 3. PVC must be Bound to a PV
+	if claim.Status.Phase != v1.ClaimBound {
+		log.WithFields(log.Fields{
+			"volumeSnapshot":  volSnap.Name,
+			"sourcePVC":       claim.Name,
+			"sourcePVCStatus": claim.Status.Phase,
+		}).Debug("Kubernetes frontend ignored this VolumeSnapshot as the source PVC is not " +
+			"bound to any volume")
+		return
+	}
+
+	// Create the volume configuration object
+	uniqueClaimName := getUniqueClaimName(claim)
+	size, _ := claim.Spec.Resources.Requests[v1.ResourceStorage]
+	accessModes := claim.Spec.AccessModes
+	annotations := claim.Annotations
+	volConfig := getVolumeConfig(accessModes, uniqueClaimName, size, annotations)
+
+	// Create the snapshot
+	snapshotExt, err := p.orchestrator.CreateVolumeSnapshot(volSnap.Name, volConfig)
+	if err != nil {
+		newVolSnap, uerr := p.updateVolumeSnapshotErrorStatus(volSnap, err.Error())
+		if uerr != nil {
+			p.eventRecorder.Event(volSnap, v1.EventTypeWarning, "UpdateStatusFailed",
+				"Snapshots controller failed to update VolumeSnapshotErrorStatus")
+			return
+		}
+		p.eventRecorder.Event(newVolSnap, v1.EventTypeNormal, "CreateSnapshotFailed", err.Error())
+		return
+	}
+
+	// Parse creation time
+	var createdAt time.Time
+	if createdAt, err = time.Parse(time.RFC3339, snapshotExt.Snapshot.Created); err != nil {
```

Comment:
[minor] why not push this down into updateVolumeSnapshotStatus. You can then just pass `volSnap` and `snapshotExt` as parameters.